core/src/Kokkos_HIP.hpp

```diff
@@ -130,6 +133,11 @@ class HIP {
     explicit SelectDevice(int id) : hip_device_id(id) {}
   };
 
+  inline uint32_t device_id() const {
+    return (static_cast<uint32_t>(device_type_id)
+            << Kokkos::Profiling::instance_bits) +
+           get_instance_id();
```

Comment:
This code is unnecessarily duplicated.  We can (and should) produce a generic version.